src/core/net/dnssd_server.cpp

```diff
@@ -107,13 +115,21 @@ void Server::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessa
 
 void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
 {
-    Error  error = kErrorNone;
     Header requestHeader;
 
-    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), requestHeader));
-    VerifyOrExit(requestHeader.GetType() == Header::kTypeQuery, error = kErrorDrop);
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+    // We first let the `Srp::Server` process the received message.
+    // It returns `kErrorNone` to indicate that it successfully
+    // processed the message.
+
+    VerifyOrExit(Get<Srp::Server>().HandleDnssdServerUdpReceive(aMessage, aMessageInfo) != kErrorNone);
```

Comment:
This may result in sending multiple responses for a single DNS query: If this message is a SRP message but the key and signature doesn't match, the SRP server will send a Security error response and here the DNS-SD server will also send a `NotImplemented` response as it is not a standard DNS query.

The `Srp::Server` returns `kErrorDrop` and silently drop this message if it is not a DNS UPDATE message, I think we should try DNS-SD server message handler only if `Get<Srp::Server>().HandleDnssdServerUdpReceive(...)` returns `kErrorDrop`, thoughts?

Reply:
> If this message is a SRP message but the key and signature doesn't match, the SRP server will send a Security error response and here the DNS-SD server will also send a NotImplemented response as it is not a standard DNS query.

@wgtdkp I don't think this can happen in the current code. Here is the `Srp::Server::ProcessMessage()`

```c++
Error Server::ProcessMessage(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
{
    Error             error;
    Dns::UpdateHeader dnsHeader;
    uint16_t          offset = aMessage.GetOffset();

    SuccessOrExit(error = aMessage.Read(offset, dnsHeader));
    offset += sizeof(dnsHeader);

    // Handles only queries.
    VerifyOrExit(dnsHeader.GetType() == Dns::UpdateHeader::Type::kTypeQuery, error = kErrorDrop);

    switch (dnsHeader.GetQueryType())
    {
    case Dns::UpdateHeader::kQueryTypeUpdate:
        ProcessDnsUpdate(aMessage, aMessageInfo, dnsHeader, offset);
        break;
    default:
        error = kErrorDrop;
        break;
    }

exit:
    return error;
}
```
If the DNS message is an query "update" then only SRP server will process it (note that no error from `ProcessDnsUpdate()` method call).

Reply:
You are right, I missed that `ProcessDnsUpdate` doesn't propagate the error back.